c++/src/Timezone.cc

```diff
@@ -735,18 +739,36 @@ namespace orc {
    * Get the local timezone.
    */
   const Timezone& getLocalTimezone() {
+#ifdef TZ_DB_STATIC_TZ
+#define TZ_STR(x) #x
+    return getTimezoneByName(TZ_STR(TZ_DB_STATIC_TZ));
+#else
     return getTimezoneByFilename(LOCAL_TIMEZONE);
+#endif
   }
 
   /**
    * Get a timezone by name (eg. America/Los_Angeles).
    * Results are cached.
    */
   const Timezone& getTimezoneByName(const std::string& zone) {
+#ifdef USE_TZ_DB
+    {
+      std::lock_guard<std::mutex> timezone_lock(timezone_mutex);
+      std::map<std::string, std::shared_ptr<Timezone> >::iterator itr =
+          timezoneCache.find(zone);
+      if (itr != timezoneCache.end()) {
+        return *(itr->second).get();
+      }
+      timezoneCache[zone] = std::shared_ptr<Timezone>(new TimezoneImpl(zone, TZ_DATABASE[zone]));
```

Comment:
why don't we pre populate the whole time zone map so you don't need to acquire a mutex every time accessing the map?